File: MyHordesOptimizerWebsite/src/app/_abstract_model/services/note.service.ts
import { HttpParams, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map, Observable } from 'rxjs';

import { NoteDTO } from '../dto/note.dto';
import { Dictionary } from '../types/_types';
import { GlobalService } from './_global.service';

@Injectable({ providedIn: 'root' })
export class NoteService extends GlobalService {

    /** Toutes les notes de ville de l'appelant, indexées par mapId. Un seul appel pour toute une liste. */
    public getMyTownNotes(): Observable<Dictionary<NoteDTO>> {
        return this.get<Dictionary<NoteDTO>>(`${this.API_URL}/Note/town/mine`, true).pipe(
            map((response: HttpResponse<Dictionary<NoteDTO>>) => response.body ?? {})
        );
    }

    public saveTownNote(mapId: number, note: string): Observable<void> {
        return this.put<void>(`${this.API_URL}/Note/town/${mapId}`, { note }).pipe(map(() => undefined));
    }

    /** Toutes les notes globales de l'appelant, indexées par userId. Un seul appel pour toute une liste. */
    public getMyUserNotes(): Observable<Dictionary<NoteDTO>> {
        return this.get<Dictionary<NoteDTO>>(`${this.API_URL}/Note/user/mine`, true).pipe(
            map((response: HttpResponse<Dictionary<NoteDTO>>) => response.body ?? {})
        );
    }

    /** Note globale (toutes villes confondues) sur un utilisateur. */
    public getUserNote(userId: number): Observable<NoteDTO> {
        return this.get<NoteDTO>(`${this.API_URL}/Note/user/${userId}`, true).pipe(
            map((response: HttpResponse<NoteDTO>) => response.body ?? { note: null })
        );
    }

    public saveUserNote(userId: number, note: string): Observable<void> {
        return this.put<void>(`${this.API_URL}/Note/user/${userId}`, { note }).pipe(map(() => undefined));
    }

    /** Toutes les notes-citoyen de l'appelant pour cette ville, indexées par userId. Un seul appel par ville. */
    public getMyCitizenNotes(townId: number): Observable<Dictionary<NoteDTO>> {
        const params: HttpParams = new HttpParams().set('townId', townId);
        return this.get<Dictionary<NoteDTO>>(`${this.API_URL}/Note/citizen/mine`, true, params).pipe(
            map((response: HttpResponse<Dictionary<NoteDTO>>) => response.body ?? {})
        );
    }

    public saveCitizenNote(userId: number, townId: number, note: string): Observable<void> {
        const params: HttpParams = new HttpParams().set('townId', townId);
        return this.put<void>(`${this.API_URL}/Note/citizen/${userId}`, { note }, params).pipe(map(() => undefined));
    }

    /** Toutes les notes-citoyen de l'appelant sur ce joueur, indexées par mapId de ville. Un seul appel pour tout son profil. */
    public getMyCitizenNotesForUser(userId: number): Observable<Dictionary<NoteDTO>> {
        return this.get<Dictionary<NoteDTO>>(`${this.API_URL}/Note/citizen/${userId}/mine`, true).pipe(
            map((response: HttpResponse<Dictionary<NoteDTO>>) => response.body ?? {})
        );
    }
}

File: MyHordesOptimizerApi/MyHordesOptimizerApi/Services/Impl/NoteService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using MyHordesOptimizerApi.Dtos.MyHordesOptimizer.Note;
using MyHordesOptimizerApi.Exceptions;
using MyHordesOptimizerApi.Models;
using MyHordesOptimizerApi.Services.Interfaces;

namespace MyHordesOptimizerApi.Services.Impl;

public class NoteService : INoteService
{
    private readonly MhoContext _dbContext;

    public NoteService(MhoContext dbContext)
    {
        _dbContext = dbContext;
    }

    /// <summary>
    /// Clé du dictionnaire = mapId (identifiant public), pas IdTown (interne) : le client envoie
    /// toujours le mapId (voir ResolveTownId), donc TownNote.IdTown doit être rejoint sur Town pour
    /// retrouver le mapId affiché par town-list. Une note sans Town.MapId (ville jamais migrée) est
    /// exclue — le front ne propose d'ailleurs pas d'ouvrir la note sur une ligne sans mapId.
    /// </summary>
    public Dictionary<int, NoteDto> GetMyTownNotes(int authorUserId)
    {
        return _dbContext.TownNotes
            .Where(n => n.IdUserAuthor == authorUserId)
            .Join(_dbContext.Towns, n => n.IdTown, t => t.IdTown,
                (n, t) => new { t.MapId, n.Note, n.UpdatedAt })
            .Where(x => x.MapId.HasValue)
            .ToDictionary(x => x.MapId!.Value, x => new NoteDto { Note = x.Note, UpdatedAt = x.UpdatedAt });
    }

    public void UpsertTownNote(int authorUserId, int mapId, string? note)
    {
        var townId = _dbContext.ResolveTownId(mapId);
        EnsureParticipated(authorUserId, townId);
        var existing = _dbContext.TownNotes
            .SingleOrDefault(n => n.IdUserAuthor == authorUserId && n.IdTown == townId);

        if (string.IsNullOrWhiteSpace(note))
        {
            if (existing is not null) _dbContext.TownNotes.Remove(existing);
            _dbContext.SaveChanges();
            return;
        }

        if (existing is null)
        {
            _dbContext.TownNotes.Add(new TownNote
            {
                IdUserAuthor = authorUserId,
                IdTown = townId,
                Note = note,
                UpdatedAt = DateTime.UtcNow
            });
        }
        else
        {
            existing.Note = note;
            existing.UpdatedAt = DateTime.UtcNow;
        }
        _dbContext.SaveChanges();
    }

    public NoteDto GetUserNote(int authorUserId, int targetUserId)
    {
        var note = _dbContext.UserNotes
            .SingleOrDefault(n => n.IdUserAuthor == authorUserId && n.IdUserTarget == targetUserId && n.IdTown == 0);
        return new NoteDto { Note = note?.Note, UpdatedAt = note?.UpdatedAt };
    }

    public void UpsertUserNote(int authorUserId, int targetUserId, string? note)
        => UpsertUserNoteInternal(authorUserId, targetUserId, 0, note);

    /// <summary>Notes globales (idTown = 0) de l'appelant, indexées par idUserTarget. Un seul appel pour toute une liste.</summary>
    public Dictionary<int, NoteDto> GetMyUserNotes(int authorUserId)
    {
        return _dbContext.UserNotes
            .Where(n => n.IdUserAuthor == authorUserId && n.IdTown == 0)
            .ToDictionary(n => n.IdUserTarget, n => new NoteDto { Note = n.Note, UpdatedAt = n.UpdatedAt });
    }

    public Dictionary<int, NoteDto> GetMyCitizenNotes(int authorUserId, int mapId)
    {
        var townId = _dbContext.ResolveTownId(mapId);
        return _dbContext.UserNotes
            .Where(n => n.IdUserAuthor == authorUserId && n.IdTown == townId)
            .ToDictionary(n => n.IdUserTarget, n => new NoteDto { Note = n.Note, UpdatedAt = n.UpdatedAt });
    }

    public void UpsertCitizenNote(int authorUserId, int targetUserId, int mapId, string? note)
    {
        var townId = _dbContext.ResolveTownId(mapId);
        EnsureParticipated(authorUserId, townId);
        UpsertUserNoteInternal(authorUserId, targetUserId, townId, note);
    }

    /// <summary>Notes de l'appelant sur ce citoyen, une par ville jouée ensemble, indexées par mapId (même join que GetMyTownNotes).</summary>
    public Dictionary<int, NoteDto> GetMyCitizenNotesForUser(int authorUserId, int targetUserId)
    {
        return _dbContext.UserNotes
            .Where(n => n.IdUserAuthor == authorUserId && n.IdUserTarget == targetUserId && n.IdTown != 0)
            .Join(_dbContext.Towns, n => n.IdTown, t => t.IdTown,
                (n, t) => new { t.MapId, n.Note, n.UpdatedAt })
            .Where(x => x.MapId.HasValue)
            .ToDictionary(x => x.MapId!.Value, x => new NoteDto { Note = x.Note, UpdatedAt = x.UpdatedAt });
    }

    private void UpsertUserNoteInternal(int authorUserId, int targetUserId, int townId, string? note)
    {
        if (authorUserId == targetUserId)
        {
            throw new MhoTechnicalException("Impossible de mettre une note sur soi-même.");
        }

        if (!_dbContext.Users.Any(u => u.IdUser == targetUserId))
        {
            throw new MhoTechnicalException($"L'utilisateur cible {targetUserId} n'existe pas.");
        }

        var existing = _dbContext.UserNotes
            .SingleOrDefault(n => n.IdUserAuthor == authorUserId && n.IdUserTarget == targetUserId && n.IdTown == townId);

        if (string.IsNullOrWhiteSpace(note))
        {
            if (existing is not null) _dbContext.UserNotes.Remove(existing);
            _dbContext.SaveChanges();
            return;
        }

        if (existing is null)
        {
            _dbContext.UserNotes.Add(new UserNote
            {
                IdUserAuthor = authorUserId,
                IdUserTarget = targetUserId,
                IdTown = townId,
                Note = note,
                UpdatedAt = DateTime.UtcNow
            });
        }
        else
        {
            existing.Note = note;
            existing.UpdatedAt = DateTime.UtcNow;
        }
        _dbContext.SaveChanges();
    }

    /// <summary>Note de ville ou de citoyen : n'a de sens que sur une ville où l'auteur a lui-même été citoyen.</summary>
    private void EnsureParticipated(int authorUserId, int townId)
    {
        var participated = _dbContext.TownCitizens.Any(c => c.IdTown == townId && c.IdUser == authorUserId);
        if (!participated)
        {
            throw new MhoTechnicalException($"L'utilisateur {authorUserId} n'a pas participé à la ville {townId}.");
        }
    }
}
